src/tools/ScenarioMeasurement/Startup/LinuxTraceSession.cs

```diff
@@ -0,0 +1,65 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace ScenarioMeasurement
+{
+    class LinuxTraceSession : ITraceSession
+    {
+        private PerfCollect perfCollect;
+        private Dictionary<TraceSessionManager.KernelKeyword, PerfCollect.KernelKeyword> linKwMapKernel;
+        private Dictionary<TraceSessionManager.ClrKeyword, PerfCollect.ClrKeyword> linKwMapClr;
+
+        public LinuxTraceSession(string sessionName, string traceName, string traceDirectory, Logger logger)
+        {
+            perfCollect = new PerfCollect(traceName, traceDirectory, logger);
+            InitLinuxKeywordMaps();
+        }
+
+        public void EnableProviders(IParser parser)
+        {
+            // Enable both providers and start the session
+            parser.EnableKernelProvider(this);
+            parser.EnableUserProviders(this);
+            perfCollect.Start();
+        }
+
+        public void Dispose()
+        {
+            perfCollect.Stop();
+        }
+
+        public void EnableKernelProvider(params TraceSessionManager.KernelKeyword[] keywords)
+        {
+            foreach (var keyword in keywords)
+            {
+                perfCollect.AddKernelKeyword(linKwMapKernel[keyword]);
+            }
+        }
+
+        public void EnableUserProvider(params TraceSessionManager.ClrKeyword[] keywords)
+        {
+            foreach (var keyword in keywords)
+            {
+                perfCollect.AddClrKeyword(linKwMapClr[keyword]);
+            }
+        }
+
+        private void InitLinuxKeywordMaps()
+        {
+            // initialize linux kernel keyword map
+            linKwMapKernel = new Dictionary<TraceSessionManager.KernelKeyword, PerfCollect.KernelKeyword>();
+            linKwMapKernel[TraceSessionManager.KernelKeyword.Process] = PerfCollect.KernelKeyword.ProcessLifetime;
+            linKwMapKernel[TraceSessionManager.KernelKeyword.Thread] = PerfCollect.KernelKeyword.Thread;
+            linKwMapKernel[TraceSessionManager.KernelKeyword.ContextSwitch] = PerfCollect.KernelKeyword.ContextSwitch;
+
+            // initialize linux clr keyword map
+            linKwMapClr = new Dictionary<TraceSessionManager.ClrKeyword, PerfCollect.ClrKeyword>();
+            linKwMapClr[TraceSessionManager.ClrKeyword.Startup] = PerfCollect.ClrKeyword.DotNETRuntimePrivate_StartupKeyword;
+        }
+
+        public void EnableUserProvider(string provider)
+        {
+        }
```

Comment:
How does enabling an `EventSource` work on Linux? I recall there's something about them getting funneled through the CLR provider somehow but I'm not sure what enables them. @brianrob may have advice here.

Reply:
My understanding is, you set the environment variable ```COMPlus_PerfMapEnabled=1``` and ```COMPlus_EnableEventLog=1``` to enable tracing config in CoreCLR. And enable the corresponding user-space events in lttng, which could be specified as command line arguments. I assume the provider information is already taken care by coreclr, no need to specify it in the tool. @brianrob  correct me if I'm wrong.

Reply:
For `EventSources` specifically, there are a few things that you need to know:

1. If you're not capturing `perf` profiles, you don't need to set `COMPlus_PerfMapEnabled=1`.  This will just do some extra work that you won't need.
2. You do need to set `COMPlus_EnableEventLog=1`.  This turns on the ability for the runtime to emit events to `LTTng`.
3. `EventSource` events are special when it comes to `LTTng`, because `LTTng` doesn't allow for new `providers` or `tracepoints` to be added on-the-fly.  Instead they all must be added at compile time.  So the runtime adds a `tracepoint` called `DotNETRuntime:EventSource` that is the single `tracepoint` for all `EventSource` events.  This also means that if you turn it on, then every `EventSource` event is enabled.  To filter this, there are a couple of environment variable you can set: `COMPlus_EventSourceFilter` will filter by `EventSource` name and `COMPlus_EventNameFilter` will filter by event name.  See https://github.com/dotnet/runtime/blob/70bfd2ac9b844701c361b90b90ef399ae6fec483/src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/TraceLogging/XplatEventLogger.cs#L162 and https://github.com/dotnet/runtime/blob/70bfd2ac9b844701c361b90b90ef399ae6fec483/src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/TraceLogging/XplatEventLogger.cs#L177 for how these are implemented.
4. Once you're able to collect these events, they will show up in `TraceEvent` with an event name of `EventSource` and the payload is JSON.  One thing that would be really nice is to have `TraceEvent` do the conversion so that they look like every other event in the system.  Note that I'm not 100% sure how complicated this will be, as there isn't a manifest, so we'd basically have to make up per-event metadata on-the-fly.  I completely understand if this turns out to be too much work and you want to punt on it.

Reply:
I would like to know the plan to get this implemented. I know that most of our current tests don't use event sources, but some of them do/will.

Reply:
Let me clarify, you mean ultimately we'd like to enable other custom event sources just like what we do in [GenericStartupParser](https://github.com/dotnet/performance/blob/afdf541e36060c0e38af3bd7737a9c977be09b71/src/tools/ScenarioMeasurement/Startup/GenericStartupParser.cs#L11-L16) ? I think this suggests we need to make significant changes to perfcollect. Currently it only allows Kernel and Clr providers.

Reply:
Resolved offline. "Punting" this issue for now.